pdal/pio.py

```diff
@@ -0,0 +1,108 @@
+import types
+import json
+import subprocess
+from functools import partial
+from collections import defaultdict
+from itertools import chain
+import copy
+import warnings
+
+import pdal
+
+try:
+    PDAL_DRIVERS_JSON = subprocess.run(["pdal", "--drivers", "--showjson"], capture_output=True).stdout
+    PDAL_DRIVERS = json.loads(PDAL_DRIVERS_JSON)
+    _PDAL_VALIDATE = True
+except:
+    PDAL_DRIVERS = []
+    _PDAL_VALIDATE = False
+
+DEFAULT_STAGE_PARAMS = defaultdict(dict)
+DEFAULT_STAGE_PARAMS.update({
+# TODO: add stage specific default configurations
+})
+
+
+class StageSpec(object):
+    def __init__(self, prefix, **kwargs):
+        self.prefix = prefix
+        self.key = ".".join([self.prefix, kwargs.get("type", "")])
+        self.spec = DEFAULT_STAGE_PARAMS[self.key].copy()
+        self.spec.update(kwargs)
+        self.spec["type"] = self.key
+        # NOTE: special case to support reading files without passing an explicit reader
+        if (self.prefix == "readers") and kwargs.get("type") == "auto":
```

Comment:
Seems like this was in the test but isn't anymore.  I don't know that you need to do anything special with this at all.  Text strings can be placed directly in the pipeline and the processor figures out what's a reader and a writer.  So I'm not sure that it really matters if you have them be a reader or writer or whatever.

Reply:
Ah yeah I had this in a test that included the categories.  I can expand this to include writers for sure, but filters should still be excluded, correct?

I didn't really want to mess with default string concatenation via `+`, or having special case module vs callable behavior so went with this:

`pipeline = pio.readers.auto("myfile.las") + ...`

over 

`pipeline = pio.readers("myfile.las") + ...`

or 

`pipeline = "myfile.las" + ...`

Reply:
I don't have strong feelings on this.  Actually, the last example seems most like what you're coercing things into, but that can be for another day.

Yes, filters are excluded, but you should be able to use "auto" with writers as well as readers.

I was just looking for a test of the "auto" stuff since it's handled specially.

Reply:
cool, yeah for sure, will get that in this morning.

Reply:
Some test cases are in and added support for auto on writers as well.

Reply:
@abellgithub Do you think this can this be resolved based on the added tests or are there any other bits I need to add?

Reply:
It's all good.  Will merge.

Reply:
@hobu: I'm wondering if this belongs in the directory that it's in right now.  Thoughts?  Also wondering about documentation.

Reply:
> Also wondering about documentation.

We don't need lots of docs at this time to merge this. If you could touch up the docstrings that should give people some hints, but we don't need a full treatise.